### ISSUE TITLE
debian/ubuntu package

### ISSUE THREAD

**Issue by chrisvire** [readbeyond/aeneas#20]
We would like to include aeneas as a package dependency on the linux version of Scripture App Builder (http://software.sil.org/scriptureappbuilder) which free software.  Is anyone working on a debian/ubuntu package?  Would you accept a pull request if I did the work as a native package or I could create a non-native package and have it in a separate repo.  What would you prefer?

**Comment by readbeyond** [comment]
On 10/26/2015 03:46 PM, Chris Hubbard wrote:
> We would like to include aeneas as a package dependency on the linux
> version of Scripture App Builder
> (http://software.sil.org/scriptureappbuilder) which free software. Is
> anyone working on a debian/ubuntu package? Would you accept a pull
> request if I did the work as a native package or I could create a
> non-native package and have it in a separate repo. What would you prefer?

Hi Chris,

I would _love_ to see a Debian/Ubuntu package! (And no, I am not working 
on it.)

I think that aeneas should be a "non-native package", according to the 
Debian FAQs: 
https://wiki.debian.org/DebianMentorsFaq#What_is_the_difference_between_a_native_Debian_package_and_a_non-native_package.3F

I plan to release aeneas v1.3.1 on 2015-10-28 (next Wed). After that, I 
want to upload aeneas to PyPI. The current setup.py should be OK already 
--- I just need some calm time to test the upload.

I am not familiar with .deb packaging: do you think having it on PyPI 
would help packaging it for Debian/Ubuntu?

AP

**Comment by chrisvire** [comment]
Here is a good discussion about this problem (i.e. a program dependent on PyPi modules:
http://askubuntu.com/questions/327543/how-can-a-debian-package-install-python-modules-from-pypi

I had a similar issue with creating the [android-sdk-installer](https://github.com/sillsdev/android-sdk-installer).  It had to download the android sdk zip files (we were not allowed to include the files in our installer due to licensing).  Based on the article above, this was a big no-no, but our hands were tied.  

So I could see an Aeneas package depending on Python and PIP and then executing pip to install the required packages.  It isn't ideal, but I think it would work well.  I will ask out packaging gurus here within SIL.  We are also looking at hosting Precise and Trusty builds of ffmpeg packages (and dependencies) at our package repository (packages.sil.org) so that all the necessary dependencies can be fulfilled.  This package repository is where we publish scripture-app-builder and android-sdk-installer packages.

**Comment by readbeyond** [comment]
On 10/26/2015 04:15 PM, Chris Hubbard wrote:
> Here is a good discussion about this problem (i.e. a program dependent
> on PyPi modules:
> http://askubuntu.com/questions/327543/how-can-a-debian-package-install-python-modules-from-pypi

I see, the security concerns are quite reasonable.

Let me list the aeneas dependecies:

1. ffmpeg (+ provides ffprobe) => currently from deb-multimedia.org, but 
soon ffmpeg should be back in the official Debian repo, or you can go 
with your own packages
2. espeak => present in the official Debian repo
3. other libs (see install_dependencies.sh) => present in the official 
Debian repo, possibly in nonfree
4. Python modules lxml, BeautifulSoup, and numpy => present in the 
official Debian repo (python-lxml, python-beautifulsoup, python-numpy)
5. Python module pafy => not available as a Debian package for Python 
2.x (there is a python3-pafy package, though); however, it is only used 
to download the audio from YouTube, and there is a safety test before 
importing it in the Python code. (I am now considering removing it from 
the requirements.txt altogether.)

 From the above, it looks like the only missing "ingredient" is the 
aeneas source code, that you could provide as part of the src of a 
native package.

Hence, please regard my above comment about a non-native package as a 
"suggestion", not a requirement. If producing a native package is 
simpler (as it seems to me now), feel free to proceed that way.

In any case, I am going to upload aeneas to PyPI too.

AP

**Comment by chrisvire** [comment]
I am looking at install_dependencies.sh.  It seems like "apt-get install -y espeak* gets way too many packages.  Which specific packages are you needing?

**Comment by readbeyond** [comment]
On 10/26/2015 09:17 PM, Chris Hubbard wrote:
> I am looking at install_dependencies.sh. It seems like "apt-get install
> -y espeak* gets way too many packages. Which specific packages are you
> needing?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/readbeyond/aeneas/issues/20#issuecomment-151272176>.
>

espeak
espeak-data

Note that install_dependencies.sh also installs other packages (vim, 
screen, etc.) as it comes from the aeneas-vagrant project, and I have 
never "pruned" it. I am opening a separate issue for that.

AP

**Comment by chrisvire** [comment]
For doing a debian non-native package, it would be helpful if there was a release tarball.  I can get a release tarball for master by accessing:
wget https://github.com/readbeyond/aeneas/tarball/master

If you were to tag your releases (in https://github.com/readbeyond/aeneas/releases), then I could access a specific release by accessing:
wget https://github.com/readbeyond/aeneas/tarball/v1.3.1

It would be a decent idea to use the release tagging. :-)

**Comment by chrisvire** [comment]
As for the pruning, If you just let me know which are not essential in a comment here, I can specify the essential dependencies in my debian/control file.

**Comment by readbeyond** [comment]
On 10/26/2015 09:55 PM, Chris Hubbard wrote:
> As for the pruning, If you just let me know which are not essential in a
> comment here, I can specify the essential dependencies in my
> debian/control file.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/readbeyond/aeneas/issues/20#issuecomment-151282320>.
>

Release tags: I will start doing it with v1.3.1. (I will try to 
reconstruct previous v1.0.0, v1.1.0, v1.2.0, v1.3.0 as well.)

Dependencies:

espeak
espeak-data
ffmpeg
flac (**)
libasound2-dev
libsndfile1-dev
libxml2-dev
libxslt-dev (*)
python-beautifulsoup
python-dev
python-lxml
python-numpy
vorbis-tools (**)

(*) not sure, it might get used to speed lxml up
(**) needed only if one reads FLAC and Vorbis files

AP

**Comment by readbeyond** [comment]
Added tags and releases

**Comment by chrisvire** [comment]
python-pip and python-scipy are not in your list above, but are in the install_dependencies and are not installed by default.  Are those needed as well?

**Comment by readbeyond** [comment]
On 10/27/2015 02:49 PM, Chris Hubbard wrote:
> python-pip and python-scipy are not in your list above, but are in the
> install_dependencies and are not installed by default. Are those needed
> as well?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/readbeyond/aeneas/issues/20#issuecomment-151502006>.
>

python-scipy is not needed in any case.

python-pip (which provides pip on $PATH) is needed only if you are going 
with the non-native option, since that will require you to pip install 
the Python modules. If you are going with the native Debian package way, 
it is not needed either.

As mentioned above, I tested the list from three comments above by 
installing those packages on a vanilla Debian jessie, and they were 
enough to have aeneas compiling and running.

(I have just rewritten install_dependencies.sh to clear the situation 
up. You can see it here: 
https://github.com/pettarin/aeneas/blob/devel/install_dependencies.sh )

**Comment by readbeyond** [comment]
Now the "pruned" list of dependencies is on both the devel/ and the master/ branches here (ReadBeyond's).

**Comment by chrisvire** [comment]
Are you wanting this package to be in the official debian/ubuntu package repository?  SIL has its own package repository (and that would be fine for us), but I don't think it serves you very well.  I assume that you want it in an official package repository.  Due to the dependency on ffmpeg, then it might be limited to vivid+ (15.04).  

After research yesterday, it seems like the correct way to get aeneas (as a Python module) into Debian/Ubuntu repositories is to work through the Debian Python Modules Team (https://wiki.ubuntu.com/Debian/PythonModulesTeam).  The process includes:
1) Create an alioth account (Gforge) at http://alioth.debian.org/
2) Ask to be added to the Debian Python Modules Team (via email or irc channel)
3) File an Intent to Package (ITP) by filing a bug report against wnpp
4) Make an account on mentors.debian.net (to upload packages for a sponsor to review)
5) use python-stdeb to develop debian subdirectory and then commit it to the *packagename*/trunk directory of the team svn repository
6) debsign package and upload it (to either mentors or sponsors)
7) ask on IRC for a Debian Developer (DD) to review it
8) after successful reviewed, upload and tag the final debian directory in svn
9) watch the NEW queue for status changes

Let me know how you want to proceed whether you want just a process to create a binary debian package (that can be hosted wherever (we would host our builds in packages.sil.org or you could do a PPA) or whether you want to go through the process of submitting to debian/ubuntu.  I will probably start as a non-native package in our git repository.  I hope to spend some time with one of our linux gurus to pair on it. 

From what I tried so far, the python-stdeb tools do a lot of the heavy lifting for creating the debian source package.

**Comment by readbeyond** [comment]
On 10/28/2015 02:30 PM, Chris Hubbard wrote:
> Are you wanting this package to be in the official debian/ubuntu package
> repository? SIL has its own package repository (and that would be fine
> for us), but I don't think it serves you very well. I assume that you
> want it in an official package repository.

Thank you for detailing the process. If you (SIL) are going to create 
and host a .deb package, that is absolutely fine with me.

At the moment, having aeneas on the public Debian/Ubuntu repos does not 
seem a top priority for the users. If that changes over time and user 
start asking for it, I will go through the official process later. 
Moreover, I expect that publishing aeneas to PyPI (which is quite 
simpler) will reduce the requests even further.

To recap: if you (SIL) want create a .deb to be hosted in your repo, 
please feel free to go ahead.

**Comment by chrisvire** [comment]
Thanks for your response.  That does simplify the process a bit. :-)  

Did you use setuptools for distutils to create your setup.py?  I am new to Python so I am trying to learn the correct way to do packaging...

**Comment by readbeyond** [comment]
On 10/28/2015 03:34 PM, Chris Hubbard wrote:
> Did you use setuptools for distutils to create your setup.py? I am new
> to Python so I am trying to learn the correct way to do packaging...

I wrote it by hand, following 
https://docs.python.org/2/distutils/setupscript.html

But, again, I am not an expert either.

In case it helps: I plan to work on uploading aeneas to PyPI next Monday 
(2015-11-02).

**Comment by chrisvire** [comment]
What is your opinion/plans in regards to Python 2 vs. Python 3?  (Sorry, I am new at Python--I only had slight understanding that there was a difference!)

There are some implications to 2 vs. 3 in packaging.  I know that Ubuntu is planning to be Python 3 by default (Python 2 still available).  I don't know of the trade-offs for you as the module developer.  Have all your dependent packages been upgraded to Python 3?

**Comment by readbeyond** [comment]
On 10/28/2015 04:51 PM, Chris Hubbard wrote:
> What is your opinion/plans in regards to Python 2 vs. Python 3? (Sorry,
> I am new at Python--I only had slight understanding that there was a
> difference!)

Python 2 and Python 3 have lots of differences, but the main one from a 
programmer's perspective consists in how strings are managed, and that 
is why it is taking forever for developers to switch to Python 3. (See 
e.g. http://nedbatchelder.com/text/unipain.html )

Now python.org encourages writing new software with Python 3.

aeneas is Python 2 because when I started working on it, the port of 
lxml to Python 3 was incomplete/buggy. Moreover, Python 2 was (and still 
is) the "default" in Debian --- which is my OS of choice.

> There are some implications to 2 vs. 3 in packaging. I know that Ubuntu
> is planning to be Python 3 by default (Python 2 still available). I
> don't know of the trade-offs for you as the module developer. Have all
> your dependent packages been upgraded to Python 3?

I have been thinking about this issue for a while. The Python 
dependencies (numpy, lxml and BeautifulSoup) are all available for 
Python 3 now (both on PyPI and Debian repos).

Porting the code of aeneas to Python 3 should not require much time --- 
internally aeneas always uses the Python 2.x "unicode" type (which will 
become Python 3.x "str" type) to handle the user-provided text, 
decoding/encoding bytes from/to UTF-8 only at the I/O interfaces. Hence, 
only the I/O interfaces need changes, and that should be quickly done.

The only problem with porting aeneas to Python 3 is the following dilemma:

- if I want to have both a Python 2.x and a Python 3.x implementation, I 
will need to maintain two (essentially identical) code bases --- and of 
course that is bad bad bad (the tools to "autoconvert" 2=>3 do not 
really work that well...);

- if I want to keep one code base only, that will force (current) users 
to switch to Python 3.x --- bad, but perhaps slightly less bad.

Given the current circumstances, it is safe to assume that aeneas will 
be available as Python 2.x only, for the next 2-3 months. Meanwhile, I 
will think/discuss on the Google Group about porting it to Python 3.x.

**Comment by chrisvire** [comment]
Have you looked at supporting Python 2 and 3 in the same codebase?

http://python3porting.com/noconv.html

**Comment by readbeyond** [comment]
On 10/28/2015 06:28 PM, Chris Hubbard wrote:
> Have you looked at supporting Python 2 and 3 in the same codebase?
>
> http://python3porting.com/noconv.html

Yes, and it looks a bit ugly to me. I would rather switch to Python 3.x 
altogether. I will ask on the aeneas Google Group what people think 
about it.

**Comment by chrisvire** [comment]
There is a python-pafy package for vivid, but not trusty.  We try to support to support all LTS release for existing software and the latest LTS for new software that we package.  We were able to get ffmpeg packages into our trusty package repository.  aeneas native package won't build on trusty due to dependency on pafy.  I really don't want to try to run pip inside package build (big no-no to do network access -- actually, I thing that the pybuild scripts that I am using setup a HTTP proxy to prevent network access during a build).  

Are you still considering removing the pafy dependency?

**Comment by readbeyond** [comment]
On 10/30/2015 07:52 PM, Chris Hubbard wrote:
> There is a python-pafy package for vivid, but not trusty. We try to
> support to support all LTS release for existing software and the latest
> LTS for new software that we package. We were able to get ffmpeg
> packages into our trusty package repository. aeneas native package won't
> build on trusty due to dependency on pafy. I really don't want to try to
> run pip inside package build (big no-no to do network access --
> actually, I thing that the pybuild scripts that I am using setup a HTTP
> proxy to prevent network access during a build).
>
> Are you still considering removing the pafy dependency?

I think we can safely consider pafy NOT as a dependency. It is used only 
in downloader.py, and there is an import check there --- so, if a user 
does not have pafy on her system, she will get a suitable message, 
prompting her to install it.

Do you need "pafy" to be removed from the setup.py and/or 
requirements.txt files to build your .deb?

AP

**Comment by chrisvire** [comment]
I get this message during building the package.  It does complete.  I will see if it works anyway:

I: dh_python2 pydist:181: Cannot find installed package that provides pafy. Using python-pafy as package name. Please add "pafy correct_package_name" line to debian/py3dist-overrides to override it IF this is incorrect.

**Comment by readbeyond** [comment]
On 10/30/2015 08:55 PM, Chris Hubbard wrote:
> I get this message during building the package. It does complete. I will
> see if it works anyway:
>
> I: dh_python2 pydist:181: Cannot find installed package that provides
> pafy. Using python-pafy as package name. Please add "pafy
> correct_package_name" line to debian/py3dist-overrides to override it IF
> this is incorrect.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/readbeyond/aeneas/issues/20#issuecomment-152636405>.
>

Just for testing this now, try removing "pafy" from setup.py and 
requirements.txt in your local copy. Now your .deb should build without 
that warning.

If so, let me know, I will do a quick update of the GitHub repo.

**Comment by chrisvire** [comment]
I am trying.  I have to fix some other things (and now I have to go home for the weekend).  I will let you know.  Thanks!

**Comment by chrisvire** [comment]
I was able to get by without pafy. Currently, the package install the contents of github.com/readbeyond/aeneas/aeneas to /usr/lib/python2.7/dist-packages/aeneas.  I noticed that the .so output file names are different ({cdtw,cew,cmfcc}.x86_64-linux-gnu.so instead of {cdtw,cew,cmfcc}.so.

If I try to run the check_dependencies script, the first 4 steps succeed, but the last 3 fail.  I tried creating a symlink and copying the file.  These still failed.  Sorry, I don't know how Python is supposed to be loading these .so files.  I am not sure what I need to do so that the .so file can be found...

**Comment by chrisvire** [comment]
OK.  This is very odd.  I am testing this inside a vagrant Trusty 64-bit machine ( trusty64 in http://github.com/chrisvire/vagrant-sil).  This configuration automatically mounts the user's home directory as ~/host in the VM.  I did "cd ~/host/src/aeneas" and "python check_dependencies.py" and loading the .so files would fail.  

However, if I "cd ~" and then run "python -v" and "import aeneas.cdtw", then the .so files are loaded just fine.

If I coipy the check_dependencies.py file to the home directory and then run it, it fails on finding ffprobe even though it is in the PATH (and I can run it from the command line).

These are some inconsistencies that I can't reconcile.

**Comment by chrisvire** [comment]
OK. I think I got everything figured out.
~/host/src/aeneas failed since it was loading the scripts local files and the libraries weren't built there.
running check_dependencies.py  from the home directory failed since it was trying to load the mp3 file out of aeneas/tests/...

So now I need to check on how Scripture App Builder is using it.
.

**Comment by chrisvire** [comment]
I have everything working now.  I will submit a PR.  I will include a commit in the PR for the removing of pafy.

**Comment by readbeyond** [comment]
Sorry for replying late, I was out of town.

I merged your PR, thank you.

About check_dependencies.py: it is outside the aeneas/ module because it was meant to be a quick way for people to detect problems with the (manual) installation/setup procedure.

I would like to keep check_dependencies.py in the GitHub repo (so that people directly cloning the repo could still use it as intended), but e.g. I will not include check_dependencies.py in the PyPI package.

=== === ===

Do you need/prefer to have the current devel/ branch merged into the master/ ?

**Comment by chrisvire** [comment]
Ok. I included check_dependencies.py in the Debian package (installed to /usr/share/python-aeneas). 

Yes, please merge with master.

**Comment by readbeyond** [comment]
Merged now. I cannot do it right now (I'm out), but I will add you to the ack section of the README as soon as possible.

**Comment by chrisvire** [comment]
FYI, you might want to review the debian/copyright file.  I think I documented the copyright correctly.  We normally license the things we do as MIT so I applied that to the debian/* files.  I am  willing for it to be AGPL if you want it.  It might make it easier.

**Comment by readbeyond** [comment]
I am fine with having the debian/* files released under MIT:

1. the AGPLv3 and the MIT are compatible;
2. the debian/* files are more configuration files rather than code;
3. you did the job, I am fine with using your default/preferred license 
on those files.

Thank you again for preparing this!

AP

**Comment by readbeyond** [comment]
Is it a problem if I rename `check_dependencies.py` as `aeneas_test_setup.py` (and update the corresponding files in debian/ as well)? If it is going to be available system-wise, I think it is a better name.

**Comment by readbeyond** [comment]
Hi Chris, in the latest release I added `aeneas_test_setup.py`, and changed the `debian` files to include it (instead of `check_dependencies.py`). The `check_dependencies.py` is there not to break SAB, but it will be removed soon.

I am thinking of moving this check script inside the `aeneas` package: we'll lose the first check (`aeneas` module can be imported) but it will be distributed on PyPI and `deb` automatically.

BTW, do you produced the `deb` archive? When you do it, would you mind letting me know? Perhaps other people, outside SIL or SAB users, might be interested in using your `deb` packages, so I will put the pointers to your repo somewhere in the documentation, thank you.

**Comment by readbeyond** [comment]
Hi @chrisvire , on the port3 branch you can find the RC1 of aeneas v1.4.0 which runs on both Python 2.7.x and 3.4/3.5.

I am not sure whether going Python 3 will simplify things for you.

**Comment by chrisvire** [comment]
@readbeyond, as best as I can remember, I got it to work with Python 2.  pybuild (used to create the packages) can generate Python 2 and Python 3 libraries (with an additional argument in debian/rules).

**Comment by readbeyond** [comment]
@chrisvire Released v1.4.0. I also updated the deb/ files, I hope correctly. Otherwise, feel free to send me a patch. I have not had time to study the "deb process", I hope I will do it some time next week.

I guess you are okay if I initiate the process to provide python-aeneas in the official Debian repository (as soon as I am comfortable with the idea of maintaining yet another piece of free sw), right?

**Comment by chrisvire** [comment]
I talked to @rpmargetts, and we are going to continue to use 1.3.1.  

If you want the package to build a library for python3, then you need to make changes. 
debian/rules:
    dh $@ --with python2,python3 --buildsystem=pybuild

debian/control:
In Build-Depends, add:
    python3-all-dev,
    python3-setuptools, python3-numpy, python3-lxml, python3-bs4,
 
Then you create the source package by running "debuild -S" in the aeneas root directory. I had to have these python3 (and python) packages installed so that I could build the source package.

I still got an error when attempting to build the binary package with pbuilder:

I: pybuild base:170: python3.4 setup.py clean 
Traceback (most recent call last):
  File "setup.py", line 52, in <module>
    long_description=open("README.rst", "r").read(),
  File "/usr/lib/python3.4/encodings/ascii.py", line 26, in decode
    return codecs.ascii_decode(input, self.errors)[0]
UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position 20501: ordinal not in range(128)
E: pybuild pybuild:256: clean: plugin distutils failed with: exit code=1: python3.4 setup.py clean 
dh_auto_clean: pybuild --clean -i python{version} -p 3.4 --dir . returned exit code 13

**Comment by readbeyond** [comment]
On 01/15/2016 04:06 PM, Chris Hubbard wrote:
> I talked to @rpmargetts <https://github.com/rpmargetts>, and we are
> going to continue to use 1.3.1.
>
> If you want the package to build a library for python3, then you need to
> make changes.
> debian/rules:
> dh $@ --with python2,python3 --buildsystem=pybuild
>
> debian/control:
> In Build-Depends, add:
> python3-all-dev,
> python3-setuptools, python3-numpy, python3-lxml, python3-bs4,
>
> Then you create the source package by running "debuild -S" in the aeneas
> root directory. I had to have these python3 (and python) packages
> installed so that I could build the source package.

Thank you for the detailed information.

> I still got an error when attempting to build the binary package with
> pbuilder:
>
> I: pybuild base:170: python3.4 setup.py clean
> Traceback (most recent call last):
> File "setup.py", line 52, in
> long_description=open("README.rst", "r").read(),
> File "/usr/lib/python3.4/encodings/ascii.py", line 26, in decode
> return codecs.ascii_decode(input, self.errors)[0]
> UnicodeDecodeError: 'ascii' codec can't decode byte 0xc3 in position
> 20501: ordinal not in range(128)
> E: pybuild pybuild:256: clean: plugin distutils failed with: exit
> code=1: python3.4 setup.py clean
> dh_auto_clean: pybuild --clean -i python{version} -p 3.4 --dir .
> returned exit code 13


I guess it would be better to use the io.open(...) with explicit utf-8 
encoding instead of open():

```
import io
long_description=io.open("README.rst", "r", "utf-8").read(),
```

Meanwhile, I think setting/exporting `PYTHONIOENCODING=UTF-8` in your 
shell should work this issue around.

AP

**Comment by pettarin** [comment]
Closing, as the setup.py has used ``io.open()`` since v1.5.0 (at least).
